packages/eslint-plugin/src/rules/member-ordering.ts

```diff
@@ -707,34 +820,111 @@ export default util.createRule<Options, MessageIds>({
 
       // Standardize config
       let order: Order | undefined;
-      let memberTypes;
+      let memberTypes: string | MemberType[] | undefined;
+      let required: 'first' | 'last' | undefined;
+
+      const memberSets: Array<Member[]> = [];
+
+      // returns true if everything is good and false if an error was reported
+      const checkOrder = (memberSet: Member[]): boolean => {
+        const hasAlphaSort = !!(order && order !== 'as-written');
+
+        // Check order
+        if (Array.isArray(memberTypes)) {
+          const grouped = checkGroupSort(
+            memberSet,
+            memberTypes,
+            supportsModifiers,
+          );
+
+          if (grouped === null) {
+            return false;
+          }
+
+          if (hasAlphaSort) {
+            return !grouped.some(
+              groupMember =>
+                !checkAlphaSort(groupMember, order as AlphabeticalOrder),
+            );
+          }
+        } else if (hasAlphaSort) {
+          return checkAlphaSort(memberSet, order as AlphabeticalOrder);
+        }
+
+        return true;
+      };
 
       if (Array.isArray(orderConfig)) {
         memberTypes = orderConfig;
       } else {
         order = orderConfig.order;
         memberTypes = orderConfig.memberTypes;
+        required = orderConfig.required;
       }
 
-      const hasAlphaSort = !!(order && order !== 'as-written');
+      if (!checkRequiredOrder(members, required)) {
+        return;
+      }
 
-      // Check order
-      if (Array.isArray(memberTypes)) {
-        const grouped = checkGroupSort(members, memberTypes, supportsModifiers);
+      if (required === 'first') {
```

Comment:
The purpose of each of these two big blocks (`required === 'first'`, `required === 'last'`) is to separate into 1-2 groups of optional vs. required members, right? If so, these can be simplified to just get the first index where two members don't have the same optionality:

```tsx
const switchIndex = members.findIndex(
  (member, index) =>
    index && isMemberOptional(member) !== isMemberOptional(members[index - 1])
);
```

_(I don't love `switchIndex` as a name, but haven't spent the time to think of a better one :smile:)_

I think you can simplify a lot of code by using this technique:
* Checking whether `switchIndex === -1` means the code can avoid creating a `memberSets` variable or intermediate array variables. It can instead directly call `checkOrder` on array slices.
* If `switchIndex > 0` and `< members.length`, you can compare the member immediately after it to the member immediately before. That effectively replaces `checkRequiredOrder`'s logic.